server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -1,156 +1,147 @@
-           throw new RuntimeException(ee);
-       }
-   }
-
-   @Override
-   public CompletableFuture<Void> asyncFlush() {
-
-       // [AG] Discuss with @crankydillo.  Here is my understanding:
-       //   - asyncFlush is used mainly for SSE.
-       //   - The idea seems to be to send the element immediately without
-       //     waiting for the rest of the elements.  Anyway, that's what SSE is.
-       //   - But, at the same time, it is trying to not overload.  So, kind of
-       //     doing backpressure.  Only if the previous flush is complete,
-       //     then request the next element from the app.
-       //   - The backpressure mechanism is already built into Reactor Netty.  But,
-       //     the question is how to communicate that.
-       //   - Please see https://projectreactor.io/docs/netty/release/reference/index.html#_sse.
-       //     It reads `The flushing strategy is "flush after every element" emitted
-       //     by the provided Publisher`.  So, we do not need to call flush individually.
-       //
-       //   In summary, we are not controlling flushing, instead Reactor Netty does.
-       //   Also, this#asyncWrite will manage the backpressure.  I assume Reactor Netty
-       //   will backpressure if it cannot flush, and that would mean an element won't be
-       //   requested so asyncWrite won't complete until the element is pulled.  Also,
-       //   asyncFlush is called right after an asyncWrite.  So, asyncFlush looks
-       //   useless.
-
-       return CompletableFuture.completedFuture(null);
-   }
-
-   @Override
-   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
-        final CompletableFuture<Void> cf = new CompletableFuture<>();
-        if (!started) {
-            byteSink = byteSinkSupplier.get();
-            parentResponse.committed();
-            started = true;
+public class ChunkOutputStream extends AsyncOutputStream {
+
+    /**
+     * This is the {@link Mono} that we return from {@link ReactorNettyJaxrsServer.Handler#handle(HttpServerRequest,
+     * HttpServerResponse)}
+     */
+    protected final Sinks.Empty<Void> completionSink;
+
+    /**
+     * Indicates that we've starting sending the response bytes.
+     */
+    private volatile boolean started;
+
+    private final NettyOutbound nettyOutbound;
+
+    /**
+     * Used in {@link NettyOutbound#send(Publisher, Predicate)} to trigger flushing the bytes before the {@link
+     * CompletableFuture} returned from `NettyOutbound#then().toFuture()` completes.
+     */
+    private static final Predicate<ByteBuf> FLUSH_ON_EACH_WRITE = bb -> true;
+
+    private final ReactorNettyHttpResponse parentResponse;
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_CANCEL = new WriterException("Response write aborted due to cancellation");
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_DISCARD = new WriterException("Response write aborted due to discarded bytes");
```

Comment:
This seems a little strange to me. It would end up with an odd stack trace.

Reply:
i have now used a simple error to indicate 'could not complete response write'.

https://github.com/resteasy/Resteasy/blob/3c53fd8b7cae7387697dc84dbf6227ef7f4f5602/server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java#L51

Please check this and let me know if this is clear enough to indicate something went wrong during response write operation.

Reply:
> For some reason GitHub won't show me the diff for `ReactorNettyHttpResponse` so hopefully this works :)
> 
> Could we not implementation [this](https://github.com/resteasy/Resteasy/blob/ca92887376a3c61d5287a1fe500e1f8ca34862e5/server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpResponse.java#L102)? I'm not sure how critical it is.
> 
> [This](https://github.com/resteasy/Resteasy/blob/ca92887376a3c61d5287a1fe500e1f8ca34862e5/server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpResponse.java#L98) should be `headers.getAll(key).add(0, String.valueOf(value));`

I will make this change. But I am not sure about the missing implementation of `equalsIgnoreValueOrder`.

@anilgursel Please comment on this if you remember why we didn't implemented this operation.

Reply:
The `equalsIgnoreValueOrder` really might not be a bit deal, I was more curious than anything.